google/resource_kms_crypto_key.go

```diff
@@ -270,118 +270,156 @@ and all its CryptoKeyVersions will be destroyed, but it will still be present on
-	return versionTemplateSchema
+func expandKmsCryptoKeyKeyRing(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	return v, nil
 }
 
-func validateKmsCryptoKeyRotationPeriod(value interface{}, _ string) (ws []string, errors []error) {
-	period := value.(string)
-	pattern := regexp.MustCompile("^([0-9.]*\\d)s$")
-	match := pattern.FindStringSubmatch(period)
+func expandKmsCryptoKeyPurpose(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	return v, nil
+}
+
+func expandKmsCryptoKeyRotationPeriod(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	return v, nil
+}
 
-	if len(match) == 0 {
-		errors = append(errors, fmt.Errorf("Invalid rotation period format: %s", period))
-		// Cannot continue to validate because we cannot extract a number.
-		return
+func expandKmsCryptoKeyVersionTemplate(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	l := v.([]interface{})
+	if len(l) == 0 || l[0] == nil {
+		return nil, nil
 	}
+	raw := l[0]
+	original := raw.(map[string]interface{})
+	transformed := make(map[string]interface{})
 
-	number := match[1]
-	seconds, err := strconv.ParseFloat(number, 64)
+	transformedAlgorithm, err := expandKmsCryptoKeyVersionTemplateAlgorithm(original["algorithm"], d, config)
+	if err != nil {
+		return nil, err
+	} else if val := reflect.ValueOf(transformedAlgorithm); val.IsValid() && !isEmptyValue(val) {
+		transformed["algorithm"] = transformedAlgorithm
+	}
 
+	transformedProtectionLevel, err := expandKmsCryptoKeyVersionTemplateProtectionLevel(original["protection_level"], d, config)
 	if err != nil {
-		errors = append(errors, err)
-	} else {
-		if seconds < 86400.0 {
-			errors = append(errors, fmt.Errorf("Rotation period must be greater than one day"))
-		}
+		return nil, err
+	} else if val := reflect.ValueOf(transformedProtectionLevel); val.IsValid() && !isEmptyValue(val) {
+		transformed["protectionLevel"] = transformedProtectionLevel
+	}
 
-		parts := strings.Split(number, ".")
+	return transformed, nil
+}
 
-		if len(parts) > 1 && len(parts[1]) > 9 {
-			errors = append(errors, fmt.Errorf("Rotation period cannot have more than 9 fractional digits"))
-		}
-	}
+func expandKmsCryptoKeyVersionTemplateAlgorithm(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	return v, nil
+}
 
-	return
+func expandKmsCryptoKeyVersionTemplateProtectionLevel(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
+	return v, nil
 }
 
-func kmsCryptoKeyNextRotation(now time.Time, period string) (result string, err error) {
-	var duration time.Duration
+func resourceKmsCryptoKeyEncoder(d *schema.ResourceData, meta interface{}, obj map[string]interface{}) (map[string]interface{}, error) {
+	// `name` and `keyRing` are used only for path building and are not sent to the server.
+	delete(obj, "name")
+	delete(obj, "keyRing")
```

Comment:
I think I can define these as `url_param_only`, right?

Reply:
Probably! Give it a try and see if it works :)